### ISSUE TITLE
Build the training surface: Matcher rename, one-verb fit API, training components (AnyMatch-first)

### ISSUE THREAD

**Issue by fxd24** [raisesquad/langres#125]
## Why this issue exists (the reasoning)

The training program (epic #85; ranked targets in `docs/plans/20260713_training_loop_plan.md`) needs framework tooling that does not exist yet. A code-level audit (2026-07-13) established the actual state:

- **Zero neural fine-tuning anywhere** — no `transformers.Trainer`, no peft/LoRA, no torch training loop; torch/sentence-transformers are inference-only. The old deep-judge adapter is a dead `NotImplementedError` stub.
- **Four disjoint "optimize" verbs** — `module.fit(candidates, labels)`, `DSPyJudge.compile(trainset)`, `derive_threshold(scores, labels)`, `BlockerOptimizer.optimize()` — with no unifying contract.
- **The fit seam is judge-only** — no `Blocker.fit`, so learned blocking (the dedupe/Zingg hallmark) is architecturally inexpressible, not just unbuilt.
- **The flywheel loop doesn't close** — `harvest_labeled_pairs` produces id-keyed pairs, but nothing converts them into a judge's positionally-aligned `fit(candidates, labels)` call. The bridge is missing.
- **A threshold cut is not calibration** — no Platt/isotonic transform, no three-band (auto-match / clerical-review / non-match) decision.

Before building, we ran a design review: a training-DX survey of 5 ER frameworks (Splink v4, dedupe, Zingg, py_entitymatching, recordlinkage) + 6 ML frameworks (sklearn, HF Trainer/TRL, Lightning, spaCy v3, DSPy, AutoGluon), a 24-use-case learning/optimization matrix over the repo's docs + `.agent/genalysis` taxonomy, and primary-source verification of AnyMatch (arXiv:2409.04073, paper + repo code).

**The governing DX criterion (maintainer, 2026-07-13):** over-simplification that hides what is trained *adds* cognitive load. PyTorch's virtue is that you always know what you're training — you see the weights, the architecture, the model. The design restores that property inside a low-ceremony API instead of trading it away.

📄 **The full decision record is `docs/research/20260714_training_surface_design.md`** (PR #124, CI green). Read it first — this issue is the execution handle for that doc, and the doc carries the evidence, the acceptance examples, and the rejected-alternatives table.

## Decided (don't relitigate here — argue on the doc PR if something looks wrong)

1. **Vocabulary: judge / `Module` → `Matcher`.** Components named by role (`Blocker`/`Comparator`/`Matcher`/`Clusterer`); the *record* vocabulary is kept (`PairwiseJudgement`, `JudgementLog`, `LinkVerdict`, `predicted_match`). Breaking rename, explicitly OK'd — no external users yet; do it before hub method-ids ossify names.
2. **API shape: A + B as one ladder.** A = explicit components (construct each trainable, call *its* `fit` — PyTorch mode, the ground truth). B = declare-then-fit (`Resolver.fit()` runs the canonical sequence over *declared* trainables) — admissible only with three honesty devices: trainables named at construction, `describe()` pre-fit, `FitReport` post-fit. Rejected: a `resolver.training.*` staged namespace, a `TrainingRecipe` orchestration noun, an external `Trainer` object, config-file-as-entry-point.
3. **One verb: `fit`.** Config in the constructor, data in `fit`, `clone()` to keep an original, hyperparameter search as a meta-object, and **the recipe/lineage is an output** (FitReport → RunRecord), never an up-front config.
4. **Labels are pairs** — `(left_id, right_id, bool)` via `LabeledPair`s or a corrections.jsonl path. Alignment to candidates (id-join + split discipline) is the library's job.
5. **AnyMatch (T2 / #83) leads the build** once Wave-0 plumbing exists — native variant first (Qwen3-class QLoRA + in-repo sklearn miner), paper-faithful (GPT-2-124M full-FT + AutoGluon) as ablation if numbers disappoint.

## What to build — phased outcomes (not prescriptions)

**Wave 0a — the Matcher rename.** One mechanical PR: `Module`→`Matcher` ABC, class/kwarg/registry/type_name/directory sweep (`core/judges/` + `core/modules/` → `core/matchers/`), `Resolver(matcher=…)` replaces `module=`, `NoJudgeAvailableError`→`NoMatcherAvailableError`, docs/examples in the same change. Scope list: design doc §1.3. ⚠️ Coordinate with the hub design's `MethodSpec` registry (`docs/research/20260713_model_identity_and_hub.md`, v0.3 step) so the three dispatch sites aren't reworked twice.

**Wave 0b — the seams.**
- `align_pairs` — the pairs→candidates bridge (id-join on `frozenset({left_id, right_id})`, split discipline). The single most load-bearing new piece; closes the flywheel loop.
- Fit protocols beyond the matcher (a `Blocker` may opt in structurally). Makes learned blocking *expressible*; implementing a trainable blocker is later work.
- `Resolver.fit` canonical sequence + `describe()` + `FitReport`.
- `budget_usd` on paid fits, enforced through the existing SpendMonitor seam, spend reported in the FitReport.

**Wave 0c — training components.**
- Miners as **plain functions** producing `LabeledPair`s (printable, inspectable data prep): `mine_misclassified` (AnyMatch hard-positives), `sample_negatives`, `attribute_examples`, `flipped`, `mine_hard_negatives` (blocking-derived, #86). Hard-positive and hard-negative mining stay separate functions (opposite failure modes); route by consumer (fine-tune wants hard, DSPy demo pools want easy-correct).
- `ColVal` serializer (Ditto/AnyMatch `COL <val>` modes, `N/A` for missing).
- `TrainedLMMatcher` scaffold (`base=`, `method=QLoRA(…)|FullFinetune()`, `head=`, `serializer=`), fit protocol, safetensors sidecar, lazy imports, training deps dev-group-only; CI dry-run on `tiny_fixture`.
- GPU smoke: Unsloth + Qwen3-0.6B QLoRA overfits 100 pairs on the 8GB 3070 (confirms tooling, calibrates real GPU-hours).

**Wave 1 — AnyMatch (T2 / #83).** LOO harness over the registry pool; honest protocol per the training plan §4; 2 headline folds (AG, Abt-Buy); target within ~5 F1 of the paper on shared datasets, deviations named.

## Definition of done — the acceptance examples

Each item is done when its example (design doc §6) is writable as shown. The three most load-bearing:

```python
# bread-and-butter supervised fit — one call, honest report
resolver = Resolver.from_schema(Company, matcher="random_forest")
report = resolver.fit(records, pairs="corrections.jsonl")
```

```python
# a flywheel round is one re-enterable call (select → label → refit)
todo = select_for_review(resolver.judgements, n=50)
# … `langres review` CLI …
report = resolver.fit(records, pairs="corrections.jsonl")
```

```python
# AnyMatch: curation is visible data prep; no Resolver involved
pairs = concat(mine_misclassified(d, miner="random_forest", cap=400)
               + sample_negatives(d, ratio=2, seed=42) for d in pool)
pairs += attribute_examples(pool, cap=600) + flipped(pairs)
matcher = TrainedLMMatcher(base="gpt2", head="classification", serializer=ColVal(mode=1))
matcher.fit(pairs)
f1 = evaluate(matcher, get_benchmark(holdout).candidates("test"), threshold=0.5)
```

## Deliberately left open (implementer's latitude)

- Exact `FitReport` schema and rendering; `describe()` output format.
- `TrainedLMMatcher` internals: unsloth vs trl/peft; serving the checkpoint via a local OpenAI-compatible `api_base` behind `LLMMatcher` vs an in-process transformers matcher (training plan §3.3 Options A/B).
- Miner signatures fine print and module placement.
- `resolve()` return shape when `decision=Bands(…)` is declared (three-way result object).
- Whether the calibrator slot + `Bands` land in Wave 0 or after AnyMatch.
- Rename-PR mechanics (single PR vs stacked; whether the deprecated `CascadeModule` dies in it).
- Wave split into sub-issues/PRs as the implementer sees fit.

## Out of scope

`partial_fit`/online + drift; weak-supervision vote-matrix label models; collective/fused clustering; the general `Optimizer`; the hub itself (#108-design); anything the training plan already defers (§7).

## Relations

- Implements the framework-deltas half of `docs/plans/20260713_training_loop_plan.md` (§3), re-expressed per the design doc.
- Unblocks #83 (AnyMatch) — Wave 1 here *is* its execution; lands #86's mining seam v1 (Wave 0c); serves #80 Phase 2 later (`CrossEncoderMatcher` baseline).
- Under epic #85; coordinate with the identity/hub design (method-id stamps become the silver-teacher lineage vocabulary).

---
Design session: https://claude.ai/code/session_015tpVEJV2CkeFFJVi1rxgrG

**Comment by fxd24** [comment]
### Scope update (2026-07-15) — 3 items added during execution

Progress: `feat/training-surface` has A (Matcher rename), B (align_pairs/FitReport), E (serve/`model_ref`), M (`method=` seam), C (`_fit_prompt`+`describe()`) merged; **PR-F (finetune/QLoRA) open as #132** with the go/no-go GREEN — a real `peft` LoRA + `trl` SFT of SmolLM2-135M, reloaded as a `model_ref`, served in-process, evaluated: **F1 0.667 → 0.927**. 4 of the 5 DoD examples already run; example 3 (capstone) is now proven.

Three items surfaced that weren't in the original plan, now scoped:

1. **GEPA optimizer** — add `GEPA(PromptMethod)` (`kind="prompt"`, reflective/Pareto prompt evolution) alongside `Bootstrap`/`MIPRO`. Verified `dspy.GEPA` ships in our pinned `dspy 3.0.3` (no dependency bump). It's a *new capability* (needs a reflection LM + feedback metric, not a MIPRO copy) → **its own PR** in the prompt-optimize family, disjoint from finetune/capstone so it runs in parallel.

2. **`from_schema(matcher="random_forest")` shorthand** — today `from_schema` only accepts `string`/`embedding`/`zero_shot_llm`/`prompt_llm`, so DoD example 1 (`Resolver.from_schema(Company, matcher="random_forest")`) doesn't run as literally written. Small `resolver.py` change + a `[trained]`-extra-missing guard → **its own small PR**, sequenced after PR-F (both touch `resolver.py`).

3. **Public-surface wiring** — the new training symbols (`Method`/`FitReport`/`MIPRO`/`Bootstrap`/`GEPA`/`QLoRA`/`finetune`/`describe`) aren't yet in `langres`/`langres.core` `__init__`, so the examples currently need deep-module imports. This is cross-cutting *integration glue* → done **uniformly in the `feat/training-surface → main` integration PR** (lazy `__getattr__` to stay import-light), with a check that all 5 DoD examples run with the clean imports they show.

Remaining after PR-F: PR-GEPA (now), then PR-G capstone + PR-D (stretch calibrator) + PR-RF after F merges, then the integration→main PR (public-surface + example validation).

**Comment by fxd24** [comment]
## ✅ Shipped — training surface landed on `main`

Merged via **#140** (`feat/training-surface → main`, merge commit `bc2b285`, 2026-07-15). `main` now carries the whole surface. Closing this epic as done.

### What shipped — 11 CI-green + Sourcery-clean sub-PRs

| Rung | PR | Delivered |
|---|---|---|
| Vocabulary | #127 | `judge`/`Module` → **`Matcher`** (role-named components; record vocab kept; wire keys stable) |
| Data bridge | #128 | `align_pairs` (labeled-pairs → candidates, entity-disjoint split, gold-coverage) + per-role fit protocols + `FitReport` |
| Serving | #129 | `api_base` threading + in-process transformers backend + `model_ref` weightless save/load |
| Method seam | #130 | `Method` base + `fit(..., method=)` dispatch to disjoint per-kind handlers |
| Fit lifecycle | #131 | canonical `Resolver.fit` + `describe()` + DSPy-compile-under-fit |
| GEPA | #133 | reflective Genetic-Pareto optimizer (`Bootstrap`/`MIPRO`/**`GEPA`**) |
| Classical baseline | #134 | `Resolver.from_schema(matcher="random_forest")` |
| Fine-tune | #132 | `finetune()` → `model_ref` + `QLoRA` + `[finetune]` extra + GPU-seconds cost |
| Capstone (DoD) | #136 | committed `train→serve→evaluate` example + `@slow @finetune` acceptance test |
| Calibration | #135 | `Platt`/`Isotonic` `Calibrator` (score→probability), Brier/ECE reported |
| Public surface | #139 | lazy top-level re-exports; the DoD examples import cleanly; import budget stays green |

### The two load-bearing simplifications held
1. **Training = `fit` with an explicit `method=` object** — `QLoRA`/`MIPRO`/`GEPA`/`Platt` carry kind + cost + config at the call site.
2. **A fine-tuned model is a `model_ref` served through the existing `LLMMatcher`** — not a new component; inference stays one seam (in-process for eval/CI, `api_base` for served endpoints).

### DoD verified by running it, not just unit tests
- **Capstone** ran on real MPS: QLoRA fine-tune → in-process serve → evaluate (Fodors-Zagat + SmolLM2-135M), $0.
- **Fine-tune genuinely learns**: F1 0.667 → 0.927 on the go/no-go (real `peft` LoRA + `trl` SFT); the acceptance gate asserts the fine-tune **out-ranks the untrained base** on separation + ROC-AUC (an F1@0.5 floor is a dishonest gate on a 135M model).
- **Calibration measurably works**: Platt/Isotonic lower Brier (0.199→0.176/0.162) and ECE (0.155→0.077/0.000).
- **CI caught a portability bug** no local run could (bf16-on-CPU in `trl`'s `SFTConfig`) — fixed + CPU-pinned regression guard.
- Quality gates on the merged branch: **2939 passed, 98.34% coverage**; import budget green; `mkdocs --strict`, mypy, ruff clean.

### Deferred (explicitly not this pass) — tracked for follow-up
- **Thinking/reasoning-model confidence** (surfaced during review): the `confidence="logprob"` probe reads the *first* generated token, so a `<think>`-first reasoning model returns `p_yes=None`. Bounded confidence-seam gap, not a rearchitect — planned modes: `verbalized` (works on hosted frontier reasoning APIs that hide logprobs; PR-D's `Calibrator` fixes overconfidence), then `logprob_after_reasoning` (in-process / vLLM-with-reasoning-parser), plus `self_consistency` as an opt-in. **To be scoped as its own issue.**
- CPU-torch-wheel for the non-blocking `test-finetune` job (speed/cleanliness only).
- From the original plan: AnyMatch LOO reproduction, miners (`mine_hard_negatives` etc.), `ColVal` serializer, `TrainableVectorBlocker` (learned blocking — now *expressible*), `PriceBook`/effective-dated pricing, classification-head matcher, DSPy-compile paid-spend capture (#100 subtask).

Design record: `docs/research/20260714_training_surface_design.md` (#124). API surface documented in `docs/TECHNICAL_OVERVIEW.md` ("Training strategies: the `method=` seam").
